underfs/b2/src/main/java/alluxio/underfs/b2/B2UnderFileSystem.java

```diff
@@ -0,0 +1,286 @@
+import com.backblaze.b2.client.exceptions.B2Exception;
+import com.backblaze.b2.client.structures.B2AccountAuthorization;
+import com.backblaze.b2.client.structures.B2Allowed;
+import com.backblaze.b2.client.structures.B2Bucket;
+import com.backblaze.b2.client.structures.B2FileVersion;
+import com.backblaze.b2.client.structures.B2ListFileNamesRequest;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.ServiceException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * B2 {@link UnderFileSystem} implementation based on the b2-sdk-core library.
+ */
+@ThreadSafe public class B2UnderFileSystem extends ObjectUnderFileSystem {
+  private static final Logger LOG = LoggerFactory.getLogger(B2UnderFileSystem.class);
+
+  public static final String B2_SCHEME = "b2://";
+  private static final String FOLDER_SUFFIX = "/";
+  private static final String FOLDER_FILE_INDICATOR = ".bzEmpty";
+
+  private final B2StorageClient mB2StorageClient;
+  private final String mBucketName;
+  private final String mAccountId;
+  private final String mBucketId;
+  private final short mBucketMode;
+
+  /**
+   * Constructs a new instance of {@link B2UnderFileSystem}.
+   *
+   * @param alluxioURI the {@link AlluxioURI} for this UFS
+   * @param conf       the configuration for this UFS
+   * @return the created {@link B2UnderFileSystem} instance
+   * @throws ServiceException
+   */
+  public static B2UnderFileSystem createInstance(AlluxioURI alluxioURI,
+      UnderFileSystemConfiguration conf) throws ServiceException {
+    LOG.debug("initializing B2 Storage Client");
+    LOG.debug(alluxioURI.toString());
+    B2StorageClient client = B2StorageClientFactory.createDefaultFactory()
+        .create(conf.get(PropertyKey.B2_ACCESS_KEY), conf.get(PropertyKey.B2_SECRET_KEY),
+            USER_AGENT);
+
+    try {
+      B2AccountAuthorization accountAuthorization = client.getAccountAuthorization();
+      LOG.debug(accountAuthorization.toString());
+
+      B2Allowed allowed = accountAuthorization.getAllowed();
+      LOG.debug(allowed.toString());
+
+      String accountId = accountAuthorization.getAccountId();
+      List<String> capabilities = allowed.getCapabilities();
+
+      String bucketNameFromURI = UnderFileSystemUtils.getBucketName(alluxioURI);
+      LOG.debug("bucketname from uri: {}", bucketNameFromURI);
+
+      String bucketId;
+      String bucketName;
+      if (allowed.getBucketId() != null) {
+        bucketId = allowed.getBucketId();
+        bucketName = allowed.getBucketName();
+      } else {
+        B2Bucket bucket = client.getBucketOrNullByName(bucketNameFromURI);
+        LOG.debug(bucket.toString());
```

Comment:
Could improve log message.

Also, adding `toString()` on objects can be expensive. Consider
```java
LOG.debug("{}", bucket);
```

This will only call the `toString()` method of the object if debug is enabled, possibly saving CPU cycles if this has occurrences in more heavily used areas of the code.

Reply:
Thanks for the tip! I changed for similar log message structure as the above comments.